[DOC] Add dissect processor documentation (#5510)
opensearch-project/documentation-website: _ingest-pipelines/processors/dissect.md

### Key modifiers 

The `dissect` processor supports key modifiers that can change the dissection's default behavior. These modifiers are always placed to the left or right of the `%{keyname}` and are always enclosed within the `%{}`. For example, the `%{+keyname->}` modifier includes the append and right padding modifiers. Key modifiers are useful for cases such as combining multiple fields into a single line of output, creating formatted lists of data items, or aggregating values from multiple sources.  

The following table lists the key modifiers for the `dissect` processor.

Modifier | Name | Position | Example | Description |
|-----------|-----------|-----------|
`->` | Skip right padding | (far) right | `%{keyname->}` | Tells `dissect` to skip over any repeated characters to the right. For example, `%{timestamp->}` could be used to tell `dissect` to skip over any padding characters, such as two spaces or any varying character padding, that follow `timestamp`. |
`+` | Append | left | `%{keyname} %{+keyname}` | Appends two or more fields together. | 

> `+` with `/n` | Append with order | left and right | `%{+keyname}/2 %{+keyname/1}` | Appends two or more fields together in the order specified. |

Author: @msfroh: Is `/n` accurate here?

The following is an example of how the right padding would be used to extract information from the following address entries: 

```bash
New York, NY 10017
New York City, NY 10017
```

Both addresses contain the same information, but the second entry has an extra word, `City`, in the city field. The right padding modifier allows the pattern to match both of these address entries, even though they have slightly different formats.

### Append modifier (`+`)

The append modifier combines the values of two or more keys into a single output value. The values are appended from left to right. You can also specify an optional separator to be inserted between the values. 

#### Example

The following pattern extracts the values of `key1` and `key2` fields and appends them together, with a space as the separator:

`%{key1} %{key2}`

The output is:

`value1 value2`

You can also specify a custom separator using the `append_separator` parameter. For example, the following pattern uses a comma as the separator: 

`%{key1} %{key2}, append_separator => ","`

The output is:

`value1, value2`

> ### Append with order modifier (`+` and `/n`)

Author: @msfroh Thank you for this comment. Will you update the text to read accurately. I'm not technical in this area, so your expertise would greatly help. Thank you.

The append with order modifier combines the values of two or more keys into a single output value, adhering to the a specific order defines by a newline character `/n`. You have the flexibility to customize the separator that separates the appended values. the append modifier is useful for compiling multiple fields into a single formatted output line, constructing structured lists of data items, and consolidating values from various sources.  

> #### Example

Author: @msfroh Please confirm I've properly interpreted your comment and example.

---
layout: default
title: Dissect
parent: Ingest processors
nav_order: 60
---

# Dissect

> The `dissect` processor extracts values from an event and maps them to individual fields based on user-defined dissect patterns. The processor is well-suited for field extractions from log messages with a known structure.

Reviewer: ```suggestion
The `dissect` processor extracts values from a document text field and maps them to individual fields based on dissect patterns. The processor is well-suited for field extractions from log messages with a known structure. Unlinke the `grok` processor, `dissect` does not use regular expressions and has a simpler syntax.
```

The right padding modifier can be used to address this issue. By adding the right padding modifier to the pattern `%{helloworldkey->} %{worldkey}`, it will no match `Hello world` (one space), `Hello  world` (two spaces), and even `Hello          world` (ten spaces). 

The right padding modifier is used to allow for the repetition of characters following a `%{keyname->}`. The right padding modifier can be applied to any key along with any other modifiers. It should always be the rightmost modifier, for example, `%{+keyname/1->}`, `%{}`.

#### Example of usage

The following is an example of how to use a right padding modifier:

`%{name->} %{city}, %{state} %{zip}`

In this pattern, the right padding modifier `->` is applied to the `%{name}` key. This means that the `%{name}` key will match an sequence of characters, including spaces. This is useful for handling names that may contain spaces, such as "First Last".

The following is an example of how the right padding would be used to extract information from the following address entries: 

```bash
New York, NY 10017
New York City, NY 10017
```

> Both addresses contain the same information, but the second entry has an extra word, `City`, in the city field. The right padding modifier allows the pattern to match both of these address entries, even though they have slightly different formats.

Author: thank you for the example :)

### Append modifier (`+`)

The append modifier combines the values of two or more keys into a single output value. The values are appended from left to right. You can also specify an optional separator to be inserted between the values. 

#### Example of usage

> The following is an example of how to use the append modifier. In this example, the pattern extracts the values of `key1` and `key2` fields and appends them together, with a space as the separator:

Reviewer: The field name to which you are appending has to be the same for all. Remove the `key1` and `key2` example and add the following example:

```
PUT /_ingest/pipeline/dissect-test
{
  "description": "Pipeline that dissects web server logs",
  "processors": [
    {
      "dissect": {
        "field": "message",
        "pattern": "%{+address}, %{+address} %{+address}",
        "append_separator": "|"
      }
    }
  ]
}
```
Test pipeline:
```
POST _ingest/pipeline/dissect-test/_simulate
{
  "docs": [
    {
      "_index": "testindex1",
      "_id": "1",
      "_source": {
        "message": "New York, NY 10017"
      }
    }
  ]
}
```
In the response, the substrings are appended into the `address` field:

```
{
  "docs": [
    {
      "doc": {
        "_index": "testindex1",
        "_id": "1",
        "_source": {
          "address": "New York|NY|10017",
          "message": "New York, NY 10017"
        },
        "_ingest": {
          "timestamp": "2024-01-22T22:30:54.516284637Z"
        }
      }
    }
  ]
}
```

Author: Thank you for the example.

`%{key1} %{key2}`

The output is:

`value1 value2`

> You can also specify a custom separator using the `append_separator` parameter. For example, the following pattern uses a comma as the separator: 

Reviewer: This syntax is incorrect. Remove this part.

`%{key1} %{key2}, append_separator => ","`

The output is:

`value1, value2`

### Append with order modifier (`+` and `/`)

> The append with order modifier combines the values of two or more keys into a single output value based on the order specified after the `/`. You have the flexibility to customize the separator that separates the appended values. The append modifier is useful for compiling multiple fields into a single formatted output line, constructing structured lists of data items, and consolidating values from various sources.  

Reviewer: This section does not show an order modifier. Here's an example that does:

Specify reverse order for the previous example
```
PUT /_ingest/pipeline/dissect-test
{
  "description": "Pipeline that dissects web server logs",
  "processors": [
    {
      "dissect": {
        "field": "message",
        "pattern": "%{+address/3}, %{+address/2} %{+address/1}",
        "append_separator": "|"
      }
    }
  ]
}
```
Test the pipeline with the same request:
```
POST _ingest/pipeline/dissect-test/_simulate
{
  "docs": [
    {
      "_index": "testindex1",
      "_id": "1",
      "_source": {
        "message": "New York, NY 10017"
      }
    }
  ]
}
```
The substrings are appended into the `address` fields in reverse order:
```
{
  "docs": [
    {
      "doc": {
        "_index": "testindex1",
        "_id": "1",
        "_source": {
          "address": "10017|NY|New York",
          "message": "New York, NY 10017"
        },
        "_ingest": {
          "timestamp": "2024-01-22T22:38:24.305974178Z"
        }
      }
    }
  ]
}
```

#### Example of usage

The following example pipeline uses the append with order modifier. Note that the `append_separator` parameter must be defined in the processor configuration, outside of the `pattern`. It is only relevant with the `+` modifier. See the following example pipeline:

```json
PUT /_ingest/pipeline/dissect-test
{
  "description": "Pipeline that dissects web server logs",
  "processors": [
    {
      "dissect": {
        "field": "message",
        "pattern": "%{a} %{+a} %{+a}",
        "append_separator":  ","
      }
    }
  ]
}
```
{% include copy-curl.html %}

If you ingest the following example document, you'll get the response `"a":"apple,banana,coconut"`:

```json
{
  "message": "apple banana coconut"
}
```
{% include copy-curl.html %}

### Named skip key

> The named skip key modifier excludes specific matches from the final output by using an empty key `{}` or `?` modifier within the pattern. For example, the following patterns are equivalent: `%{firstName} %{lastName} %{?ignore}` and `%{firstName} %{lastName} %{}`. The named skip key modifier is useful for excluding irrelevant or unnecessary fields from the output, focusing on specific information, or streamlining the output for further processing or analysis. 

Reviewer: ```suggestion
The named skip key modifier excludes specific matches from the final output by using an empty key `{}` or `?` modifier within the pattern. For example, the following patterns are equivalent: `%{firstName} %{lastName} %{?ignore}` and `%{firstName} %{lastName} %{}`. The named skip key modifier is useful for excluding irrelevant or unnecessary fields from the output. 
```

#### Example of usage

> The following pattern uses a named skip key to exclude a field (in this case, `ignore`) from the output. You can assign a descriptive name to the empty key, for example, `%{?ignore}`, to clarify that the corresponding value should be excluded from the final result:

Reviewer: How about a full example:

```
PUT /_ingest/pipeline/dissect-test
{
  "description": "Pipeline that dissects web server logs",
  "processors": [
    {
      "dissect": {
        "field": "message",
        "pattern": "%{firstName} %{lastName} %{?ignore}"
      }
    }
  ]
}
```
Test:
```
POST _ingest/pipeline/dissect-test/_simulate
{
  "docs": [
    {
      "_index": "testindex1",
      "_id": "1",
      "_source": {
        "message": "John Doe M.D."
      }
    }
  ]
}
```
Response:
```
{
  "docs": [
    {
      "doc": {
        "_index": "testindex1",
        "_id": "1",
        "_source": {
          "firstName": "John",
          "lastName": "Doe",
          "message": "John Doe M.D."
        },
        "_ingest": {
          "timestamp": "2024-01-22T22:41:58.161475555Z"
        }
      }
    }
  ]
}
```

`%{firstName} %{lastName} %{?ignore}`

### Reference keys (`*` and `&`)

Reference keys use parsed values as key/value pairings for structured content. This can use useful when handling systems that partially log data in key/value pairs. By using reference keys, you can preserve the key/value relationship and maintain the integrity of the extracted information. 

#### Example of usage

> The following pattern uses a reference key to extract data into a structured format. In this example, `%{value}` represents the parsed value and `%{reference_key}` acts as the placeholder for the actual key:

Reviewer: Here's an example:
extract `client_ip` and two key/value pairs for the next values:
```
PUT /_ingest/pipeline/dissect-test
{
  "description": "Pipeline that dissects web server logs",
  "processors": [
    {
      "dissect": {
        "field": "message",
        "pattern": "%{client_ip} %{*a}:%{&a} %{*b}:%{&b}"
      }
    }
  ]
}
```
Test the pipeline:
```
POST _ingest/pipeline/dissect-test/_simulate
{
  "docs": [
    {
      "_index": "testindex1",
      "_id": "1",
      "_source": {
        "message": "192.168.1.10 response_code:200 response_size:3456"
      }
    }
  ]
}
```
The two key/value pairs were extracted into fields:
```
{
  "docs": [
    {
      "doc": {
        "_index": "testindex1",
        "_id": "1",
        "_source": {
          "client_ip": "192.168.1.10",
          "response_code": "200",
          "message": "192.168.1.10 response_code:200 response_size:3456",
          "response_size": "3456"
        },
        "_ingest": {
          "timestamp": "2024-01-22T22:48:51.475535635Z"
        }
      }
    }
  ]
}
```

Author: Thank you for the examples.

nav_order: 60
---

# Dissect

The `dissect` processor extracts values from a document text field and maps them to individual fields based on dissect patterns. The processor is well-suited for field extractions from log messages with a known structure. Unlike the `grok` processor, `dissect` does not use regular expressions and has a simpler syntax.

## Syntax

The following is the syntax for the `dissect` processor:

```json
{
  "dissect": {
    "field": "source_field",
    "pattern": "%{dissect_pattern}"
  }
}
```
{% include copy-curl.html %}

## Configuration parameters

The following table lists the required and optional parameters for the `dissect` processor.

Parameter | Required/Optional | Description |
|-----------|-----------|-----------|
`field`  | Required  | The name of the field containing the data to be dissected. |
`pattern` | Required | The dissect pattern used to extract data from the specified field. |
`append_separator` | Optional | The separator character or string that separates appended fields. Default is `""` (empty string).
`description`  | Optional  | A brief description of the processor.  |
`if` | Optional | A condition for running the processor. |

> `ignore_failure` | Optional | Specifies whether the processor continues execution even if it encounters errors. If set to `true`, failures are ignored. Default is `false`. |

Author: Revised to clarify that it's the processor failure that is ignored, not the error. _If set to `true`, the processor failure is ignored._

          "url": "/login",
          "response_size": "3456",
          "timestamp": "03/Nov/2023:15:20:45 +0000"
        },
        "_ingest": {
          "timestamp": "2023-11-03T22:28:32.830244044Z"
        }
      }
    }
  ]
}
```

**Step 3: Ingest a document**

The following query ingests a document into an index named `testindex1`:

```json
PUT testindex1/_doc/1?pipeline=dissect-test
{
   "message": "192.168.1.10 - - [03/Nov/2023:15:20:45 +0000] \"POST /login HTTP/1.1\" 200 3456"
}
```
{% include copy-curl.html %}

**Step 4 (Optional): Retrieve the document**

To retrieve the document, run the following query:

```json
GET testindex1/_doc/1
```
{% include copy-curl.html %}

## Dissect patterns

A dissect pattern is a way to tell `dissect` how to parse a string into a structured format. The pattern is defined by the parts of the string that you want to discard. For example, the `%{client_ip} - - [%{timestamp}]` dissect pattern parses the string `"192.168.1.10 - - [03/Nov/2023:15:20:45 +0000] \"POST /login HTTP/1.1\" 200 3456"` into the following fields:

```json
client_ip: "192.168.1.1"
@timestamp: "03/Nov/2023:15:20:45 +0000"
```

> The dissect pattern works by matching the string against a set of rules. For example, the first rule is to discard a single space. Dissect will find this space and then assign the value of `client_ip` to everything up to that space. The next rule is to match the `[` and `]` characters and then assign the value of `@timestamp` to everything in between.

Author: Revised to read "...to all the data characters before that space."

### Building successful dissect patterns

When building dissect pattern, it is important to pay attention to the parts of the string that you want to discard. If you discard too much of the string, then `dissect` may not be able to successfully parse the remaining data. Conversely, if you do not discard enough of the string, then `dissect` may create unnecessary fields.

If any `%{keyname}` defined in the pattern do not have a value, then an exception is thrown. You can handle this exception by providing error handling steps in the `on_failure` parameter.

### Empty and named skip keys

An empty key `%{}` or a [named skip key](#named-skip-key) can be used to match values but exclude the value from the final document. This can be useful if you want to parse a string but do not need to store all its parts.

### Converting matched values to a non-string data type

By default, all matched values are represented as string data types. If you need to convert a value to a different data type, you can use the [`convert` processor]({{site.url}}{{site.baseurl}}/ingest-pipelines/processors/convert/).

### Key modifiers 

> The `dissect` processor supports key modifiers that can change the default processor behavior. These modifiers are always placed to the left or right of the `%{keyname}` and are always enclosed within `%{}`. For example, the `%{+keyname->}` modifier includes the append and right padding modifiers. Key modifiers are useful for cases such as combining multiple fields into a single line of output, creating formatted lists of data items, or aggregating values from multiple sources.  

Author: They are common names and not used in the code. The names used in the code vary, such as `append_separator` or `->`, which is the right padding modifier.

> The `dissect` processor supports key modifiers that can change the default processor behavior. These modifiers are always placed to the left or right of the `%{keyname}` and are always enclosed within `%{}`. For example, the `%{+keyname->}` modifier includes the append and right padding modifiers. Key modifiers are useful for cases such as combining multiple fields into a single line of output, creating formatted lists of data items, or aggregating values from multiple sources.  

Reviewer 2: Ok, can we indicate that they're names in some way? Or maybe hyphenate "right-padding"?

Detailed descriptions of each key modifier, along with usage examples, are in the following sections.

### Right padding modifier (`->`)

The dissection algorithm is precise and requires that every character in the pattern exactly match the source string. For instance, the pattern `%{hellokey} %{worldkey}` (one space) will match the string "Hello world" (one space) but not the string "Hello  world" (two spaces) because the pattern only has one space while the source string has two.

The right padding modifier can be used to address this issue. By adding the right padding modifier to the pattern `%{helloworldkey->} %{worldkey}`, it will match <code>Hello&nbsp;world</code> (one space), <code>Hello&nbsp;&nbsp;world</code> (two spaces), and even <code>Hello&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;world</code> (ten spaces). 

The right padding modifier is used to allow for the repetition of characters following a `%{keyname->}`. The right padding modifier can be applied to any key along with any other modifiers. It should always be the rightmost modifier, for example, `%{+keyname/1->}` or `%{}`.

> #### Example of usage

Reviewer 2: ```suggestion
#### Example usage
```

The following is an example of how to use a right padding modifier:

`%{city->}, %{state} %{zip}`

> In this pattern, the right padding modifier `->` is applied to the `%{city}` key. Both addresses contain the same information, but the second entry has an extra word, `City`, in the city field. The right padding modifier allows the pattern to match both of these address entries, even though they have slightly different formats: 

Reviewer 2: ```suggestion
In this pattern, the right padding modifier `->` is applied to the `%{city}` key. Both addresses contain the same information, but the second entry has an extra word, `City`, in the city field. The right padding modifier allows the pattern to match both of these address entries even though they have slightly different formats: 
```

> In this pattern, the right padding modifier `->` is applied to the `%{city}` key. Both addresses contain the same information, but the second entry has an extra word, `City`, in the city field. The right padding modifier allows the pattern to match both of these address entries, even though they have slightly different formats: 

Reviewer 2: Should "city" be in code font?

```bash
New York, NY 10017
New York City, NY 10017
```

The following example pipeline uses the right padding modifier with an empty key `%{->}`:

```json
PUT /_ingest/pipeline/dissect-test
{
  "description": "Pipeline that dissects web server logs",
  "processors": [
    {
      "dissect": {
        "field": "message",
        "pattern": "[%{client_ip}]%{->}[%{timestamp}]" 
      }
    }
  ]
}
```
{% include copy-curl.html %}

> You can test the pipeline using the following example:

Reviewer 2: ```suggestion
You can test the pipeline by using the following example:
```

> You can test the pipeline using the following example:

Reviewer 2: Can we add the noun after "example"?

```json
POST _ingest/pipeline/dissect-test/_simulate
{
  "docs": [
    {
      "_index": "testindex1",
      "_id": "1",
      "_source": {
        "message": "[192.168.1.10]   [03/Nov/2023:15:20:45 +0000]"
      }
    }
  ]
}
```
{% include copy-curl.html %}

> Your response should be similar to the following:

Reviewer 2: ```suggestion
Your response should appear similar to the following:
```

```json
{
  "docs": [
    {
      "doc": {
        "_index": "testindex1",
        "_id": "1",
        "_source": {
          "client_ip": "192.168.1.10",
          "message": "[192.168.1.10]   [03/Nov/2023:15:20:45 +0000]",
          "timestamp": "03/Nov/2023:15:20:45 +0000"
        },
        "_ingest": {
          "timestamp": "2024-01-22T22:55:42.090569297Z"
        }
      }
    }
  ]
}
```
{% include copy-curl.html %}

### Append modifier (`+`)

> The append modifier combines the values of two or more values into a single output value. The values are appended from left to right. You can also specify an optional separator to be inserted between the values. 

Reviewer 2: Global: Should the names of the modifiers be in code font or otherwise formatted to indicate that they are names?

Reviewer 2: In that case, it should be "An append modifier".

> #### Example of usage

Reviewer 2: ```suggestion
#### Example usage
```

The following is an example pipeline with an append modifier: 

```json
PUT /_ingest/pipeline/dissect-test
{
  "description": "Pipeline that dissects web server logs",
  "processors": [
    {
      "dissect": {
        "field": "message",
        "pattern": "%{+address}, %{+address} %{+address}",
        "append_separator": "|"
      }
    }
  ]
}
```
{% include copy-curl.html %}

> You can test the pipeline using the following example:

Reviewer 2: ```suggestion
You can test the pipeline by using the following example:
```

> You can test the pipeline using the following example:

Reviewer 2: Can we add the noun after "example"?

```json
POST _ingest/pipeline/dissect-test/_simulate
{
  "docs": [
    {
      "_index": "testindex1",
      "_id": "1",
      "_source": {
        "message": "New York, NY 10017"
      }
    }
  ]
}
```
{% include copy-curl.html %}

The substrings are appended to the `address` field, as shown in the following response:

```json
{
  "docs": [
    {
      "doc": {
        "_index": "testindex1",
        "_id": "1",
        "_source": {
          "address": "New York|NY|10017",
          "message": "New York, NY 10017"
        },
        "_ingest": {
          "timestamp": "2024-01-22T22:30:54.516284637Z"
        }
      }
    }
  ]
}
```
{% include copy-curl.html %}

> ### Append with order modifier (`+` and `/`)

Reviewer: ```suggestion
### Append with order modifier (`+` and `/n`)
```

> ### Append with order modifier (`+` and `/n`)

Reviewer 2: Re: my previous comments, this modifier name is particularly awkward without some way of indicating that it's a name.

> The append with order modifier combines the values of two or more keys into a single output value based on the order specified after the `/` modifier. You have the flexibility to customize the separator that separates the appended values. The append modifier is useful for compiling multiple fields into a single formatted output line, constructing structured lists of data items, and consolidating values from various sources.  

Reviewer: ```suggestion
The append with order modifier combines the values of two or more keys into a single output value based on the order specified after the `/`. You have the flexibility to customize the separator that separates the appended values. The append modifier is useful for compiling multiple fields into a single formatted output line, constructing structured lists of data items, and consolidating values from various sources.  
```

> #### Example of usage

Reviewer 2: ```suggestion
#### Example usage
```

> The following example pipeline uses the append with order modifier to reverse the pattern order defined in the preceding pipeline. The `append_separator` parameter must be defined in the processor configuration, outside of the `pattern`. It is only relevant with the `+` modifier.

Reviewer: ```suggestion
The following example pipeline uses the append with order modifier to reverse the pattern order defined in the preceding pipeline. This pipeline specifies a separator to insert between the appended fields. If you don't specify a separator, all values will be appended together without a separator.
```

```json
PUT /_ingest/pipeline/dissect-test
{
  "description": "Pipeline that dissects web server logs",
  "processors": [
    {
      "dissect": {
        "field": "message",
        "pattern": "%{+address/3}, %{+address/2} %{+address/1}",
        "append_separator": "|"
      }
    }
  ]
}
```
{% include copy-curl.html %}

> You can test the pipeline using the following example:

Reviewer 2: Can we add the noun after "example"?

```json
POST _ingest/pipeline/dissect-test/_simulate
{
  "docs": [
    {
      "_index": "testindex1",
      "_id": "1",
      "_source": {
        "message": "New York, NY 10017"
      }
    }
  ]
}
```
{% include copy-curl.html %}

> The substrings are appended into the `address` field in reverse order, as shown in the following response:

Reviewer 2: ```suggestion
The substrings are appended to the `address` field in reverse order, as shown in the following response:
```

```json
{
  "docs": [
    {
      "doc": {
        "_index": "testindex1",
        "_id": "1",
        "_source": {
          "address": "10017|NY|New York",
          "message": "New York, NY 10017"
        },
        "_ingest": {
          "timestamp": "2024-01-22T22:38:24.305974178Z"
        }
      }
    }
  ]
}
```
{% include copy-curl.html %}

### Named skip key

> The named skip key modifier excludes specific matches from the final output by using an empty key `{}` or `?` modifier within the pattern. For example, the following patterns are equivalent: `%{firstName} %{lastName} %{?ignore}` and `%{firstName} %{lastName} %{}`. The named skip key modifier is useful for excluding irrelevant or unnecessary fields from the output.

Reviewer 2: ```suggestion
The named skip key modifier excludes specific matches from the final output by using an empty key (`{}` or `?`) modifier within the pattern. For example, the following patterns are equivalent: `%{firstName} %{lastName} %{?ignore}` and `%{firstName} %{lastName} %{}`. The named skip key modifier is useful for excluding irrelevant or unnecessary fields from the output.
```

> The named skip key modifier excludes specific matches from the final output by using an empty key `{}` or `?` modifier within the pattern. For example, the following patterns are equivalent: `%{firstName} %{lastName} %{?ignore}` and `%{firstName} %{lastName} %{}`. The named skip key modifier is useful for excluding irrelevant or unnecessary fields from the output.

Reviewer 2: This modifier name is also very awkward without some formatting to indicate that it's a name.

> #### Example of usage

Reviewer 2: ```suggestion
#### Example usage
```

> The following pattern uses a named skip key to exclude a field (in this case, `ignore`) from the output. You can assign a descriptive name to the empty key, for example, `%{?ignore}`, to clarify that the corresponding value should be excluded from the final result:

Reviewer 2: ```suggestion
The following pattern uses a named skip key to exclude a field (in this case, `ignore`) from the output. You can assign a descriptive name to the empty key, for example, `%{?ignore}`, to clarify that the corresponding value should be excluded from the final output:
```

> The following pattern uses a named skip key to exclude a field (in this case, `ignore`) from the output. You can assign a descriptive name to the empty key, for example, `%{?ignore}`, to clarify that the corresponding value should be excluded from the final result:

Reviewer 2: Should "modifier" follow "named skip key"?

```json
PUT /_ingest/pipeline/dissect-test
{
  "description": "Pipeline that dissects web server logs",
  "processors": [
    {
      "dissect": {
        "field": "message",
        "pattern": "%{firstName} %{lastName} %{?ignore}"
      }
    }
  ]
}
```
{% include copy-curl.html %}

> You can test the pipeline using the following example:

Reviewer 2: ```suggestion
You can test the pipeline by using the following example:
```

> You can test the pipeline using the following example:

Reviewer 2: Can we add the noun after "example"?

```json
POST _ingest/pipeline/dissect-test/_simulate
{
  "docs": [
    {
      "_index": "testindex1",
      "_id": "1",
      "_source": {
        "message": "John Doe M.D."
      }
    }
  ]
}
```
{% include copy-curl.html %}

> Your response should be similar to the following:

Reviewer 2: ```suggestion
Your response should appear similar to the following:
```

```json
{
  "docs": [
    {
      "doc": {
        "_index": "testindex1",
        "_id": "1",
        "_source": {
          "firstName": "John",
          "lastName": "Doe",
          "message": "John Doe M.D."
        },
        "_ingest": {
          "timestamp": "2024-01-22T22:41:58.161475555Z"
        }
      }
    }
  ]
}
```
{% include copy-curl.html %}

### Reference keys (`*` and `&`)

> Reference keys use parsed values as key/value pairings for structured content. This can use useful when handling systems that partially log data in key/value pairs. By using reference keys, you can preserve the key/value relationship and maintain the integrity of the extracted information. 

Reviewer 2: ```suggestion
Reference keys use parsed values as key-value pairings for structured content. This can use useful when handling systems that partially log data in key-value pairs. By using reference keys, you can preserve the key-value relationship and maintain the integrity of the extracted information. 
```

> #### Example of usage

Reviewer 2: ```suggestion
#### Example usage
```

> The following pattern uses a reference key to extract data into a structured format. In this example, `client_ip` and two key/value pairs are extracted for the next values:

Reviewer 2: ```suggestion
The following pattern uses a reference key to extract data into a structured format. In this example, `client_ip` and two key-value pairs are extracted for the next values:
```

> The following pattern uses a reference key to extract data into a structured format. In this example, `client_ip` and two key/value pairs are extracted for the next values:

Reviewer 2: "extract...into" doesn't work. Please revise. Do we mean something like "to extract data for use in a structured format"?

Reviewer 2: Follow-up: @kolchfa-aws has indicated that "extract into" is common usage in a CS context. I would still prefer different/clearer wording but will defer to you.

Author: Revised to read: "to find data and organize data it into a more structured format."

```json
PUT /_ingest/pipeline/dissect-test
{
  "description": "Pipeline that dissects web server logs",
  "processors": [
    {
      "dissect": {
        "field": "message",
        "pattern": "%{client_ip} %{*a}:%{&a} %{*b}:%{&b}"
      }
    }
  ]
}
```
{% include copy-curl.html %}

> You can test the pipeline using the following example:

Reviewer 2: ```suggestion
You can test the pipeline by using the following example:
```

> You can test the pipeline using the following example:

Reviewer 2: Can we add the noun after "example"?

```json
POST _ingest/pipeline/dissect-test/_simulate
{
  "docs": [
    {
      "_index": "testindex1",
      "_id": "1",
      "_source": {
        "message": "192.168.1.10 response_code:200 response_size:3456"
      }
    }
  ]
}
```
{% include copy-curl.html %}

> The two key/value pairs were extracted into fields, as shown in the following response:

Reviewer 2: ```suggestion
The two key-value pairs were extracted into fields, as shown in the following response:
```

> The two key/value pairs were extracted into fields, as shown in the following response:

Reviewer 2: "extracted into" doesn't work (we would normally use "extracted from"). Do we mean something like "were extracted and then added to fields"?

Reviewer 2: Follow-up: @kolchfa-aws has indicated that "extract into" is common usage in a CS context. I would still prefer different/clearer wording but will defer to you.

Author: Revised to read "to find data and organize it into a structured format."
